File: parallel1/ParallelSum.py
import random
import time
from concurrent.futures import ProcessPoolExecutor

def calculate_partial_sum(numbers, start_index, end_index):
    return sum(numbers[start_index:end_index])

def main():
    array_size = int(input("Enter array size: "))
    number_of_parts = int(input("Enter number of parts (processes): "))

    numbers = [random.randint(0, 99) for _ in range(array_size)]

    base_part_size = array_size // number_of_parts
    remaining_elements = array_size % number_of_parts

    indices = []
    current_index = 0
    for part_index in range(number_of_parts):
        start_index = current_index
        elements_in_part = base_part_size + (1 if part_index < remaining_elements else 0)
        end_index = start_index + elements_in_part
        indices.append((start_index, end_index))
        current_index = end_index

    with ProcessPoolExecutor(max_workers=number_of_parts) as executor:
        futures = [
            executor.submit(calculate_partial_sum, numbers, start, end)
            for start, end in indices
        ]
        total_sum = sum(f.result() for f in futures)

    print(f"Total sum of array elements: {total_sum}")

if __name__ == "__main__":
    main()

File: parallel1/ParallelSum.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;
import java.util.Random;
import java.util.Scanner;

public class ParallelSum {

    public static long calculatePartialSum(int[] array, int startIndex, int endIndex) {
        long sum = 0;
        for (int i = startIndex; i < endIndex; i++) {
            sum += array[i];
        }
        return sum;
    }

    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter array size: ");
        int arraySize = scanner.nextInt();

        System.out.print("Enter number of parts (threads): ");
        int numberOfParts = scanner.nextInt();

        scanner.close();

        int[] array = new int[arraySize];
        Random random = new Random();
        for (int i = 0; i < arraySize; i++) {
            array[i] = random.nextInt(100);
        }

        int basePartSize = arraySize / numberOfParts;
        int remainingElements = arraySize % numberOfParts;
        int currentIndex = 0;

        try (ExecutorService executor = Executors.newFixedThreadPool(numberOfParts)) {
            List<Future<Long>> futures = new ArrayList<>();

            for (int partIndex = 0; partIndex < numberOfParts; partIndex++) {
                int startIndex = currentIndex;
                int elementsInThisPart = basePartSize + (partIndex < remainingElements ? 1 : 0);
                int endIndex = startIndex + elementsInThisPart;
                futures.add(executor.submit(() -> calculatePartialSum(array, startIndex, endIndex)));
                currentIndex = endIndex;
            }

            long totalSum = 0;
            for (Future<Long> future : futures) {
                totalSum += future.get();
            }

            System.out.println("Total sum of array elements: " + totalSum);
        }
    }
}
